lib/scanner/walk.go

```diff
@@ -238,48 +273,177 @@ func (w *walker) walkAndHashFiles(ctx context.Context, fchan, dchan chan protoco
 		}
 
 		if fs.IsInternal(path) {
-			l.Debugln("ignored (internal):", path)
+			l.Debugln("skip walking (internal):", path)
 			return skip
 		}
 
 		if w.Matcher.Match(path).IsIgnored() {
-			l.Debugln("ignored (patterns):", path)
+			l.Debugln("skip walking (patterns):", path)
+			return skip
+		}
+
+		if err != nil {
+			if sendErr := fsWalkError(ctx, fsChan, path, err); sendErr != nil {
+				return sendErr
+			}
 			return skip
 		}
 
 		if !utf8.ValidString(path) {
+			if err := fsWalkError(ctx, fsChan, path, errors.New("path isn't a valid utf8 string")); err != nil {
+				return err
+			}
 			l.Warnf("File name %q is not in UTF8 encoding; skipping.", path)
 			return skip
 		}
 
 		path, shouldSkip := w.normalizePath(path, info)
 		if shouldSkip {
+			if err := fsWalkError(ctx, fsChan, path, errors.New("failed to normalize path")); err != nil {
+				return err
+			}
 			return skip
 		}
 
-		switch {
-		case info.IsSymlink():
-			if err := w.walkSymlink(ctx, path, dchan); err != nil {
-				return err
+		select {
+		case fsChan <- fsWalkResult{
+			path: path,
+			info: info,
+			err:  nil,
+		}:
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+
+		// under no circumstances shall we descend into a symlink
```

Comment:
I'm 100% sure this is mythical and can be removed

Reply:
Let me recheck this (sorry, I am a bit cloudy atm): This case already existed pre-PR. Descending into symlinked dirs should already be prevented in `fs.Walk`, so I agree this check is redundant. I understood it as extra security for a messy topic (currently I am thinking about the platform dependent behaviour of `IsDir` or the whole reparse point BS) and making it explicit, that we skip symlinks. Should this be removed (possibly adding a comment about skipping symlinks)?